### PR TITLE
PHP Sessions and Cookies to Create a Secure 'Remember Me' Section In a Login Script

### LINKED ISSUE #3063: PHP Sessions and Cookies to Create a Secure 'Remember Me' Section In a Login Script
### Proposed title of article
 PHP Sessions and Cookies to Create a Secure 'Remember Me' Section In a Login Script.
### Introduction paragraph (2-3 paragraphs):
The ‘Remember Me' functionality in the Login Script will allow the user to keep their logged in status. 
When a user selects the Remember Me option, the logged-in status is serialized in the PHP session or cookies.
While writing user login data to a session or cookie,It is critical to be alert of security problems that could jeopardize the program's authentication system.

Because plain passwords can be hacked, they should not be saved in the user's cookie.
This article will show you how to make a PHP web application with a persistent authentication system.
When a user tries to log in using the application, the entered login credentials are validated against the database.

If a match is detected, the user's logged-in status is preserved using the PHP session and cookies before being redirected to the dashboard. The unique member id from the member database is saved in a session after a successful login. The cookies are then set to save the login name and password for a given length of time. 
To prevent hacking, a random password and a token are produced and kept in the cookie instead of the users' plain password. When a user visits the application pages, the user's current logged-in session is checked and the user is redirected to the requested page. If the session is empty, the code will check the cookies to see if the user is logged in. The user will be sent back to the login page if neither the session nor the cookies include any information about the remembered login.

### Key takeaways: 
1.Create Cookies to Preserve the Logged-In State
2.Validate Remembered Login with PHP Session and Cookies
3.Clear Remembered Login with Session and Cookies on Logout

### References: 
N/A

### PR DESCRIPTION
**Pull request by Juskam**
…Login Script

### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #3063 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+Before redirecting the user to their dashboard, the PHP session and cookies are used to preserve the user's logged-in state. Session IDs are created when users successfully log in.
+The cookies are then set to save the login name and password for a given length of time. To prevent hacking, a random password and token are produced and kept in the cookie instead of the users' plain password.
+While accessing application pages, a check is made to see if the user is currently logged in, and if not, the page is not displayed. After checking if the session is empty, a cookie check will be performed to determine if the user is logged in. This happens when both the session and cookies do not contain any information about a remembered login.
+With a 1 month expiration date, authentication cookies can be used to authenticate a user. It will be possible to store random passwords and tokens in a database along with their expiration date and time. When logging in, the cookie-based logged-in state validation is performed by checking the database for cookie expiration and availability.
+
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:12]: @Juskam, this intro is too wordy. Kindly find a way to remove some content from here and add to the article body. Reason being you have covered almost everything here.

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+**Prerequisites**
+
+- PHP and MSQL
+- PHP cookies
+- PHP sessions
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:17]: What level of knowledge does the reader need to get started? Expert? Beginner? and what's difference having PHP skills and PHP session and cookies? Avoid repetition as much as possible.

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+#### Table of contents
+
+- [Introduction](#introduction)
+
+- [Preserve Logged in State](#create-cookies-to-preserve-the-logged-in-state)
+- [validate remembered login](#validate-remembered-login-with-php-session-and-cookies)
+- [Clear Remembered Login](#clear-remembered-login-with-session-and-cookies-on-logout)
+- [Controller Classes](#auth-and-dbcontroller-classes)
+- [conclusion](#conclusion)
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:27]: Inconsistent capitalization. Kindly fix this.

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+### Create Cookies to Preserve the Logged-In State
+
+---
+
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:32]: Remove these unneccessary spaces where applicable.

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+        </div>
+    </div>
+</form>
+```
+
+### Validate Remembered Login with PHP Session and Cookies
+
+---
+
+authCookieSessionValidate.php, a PHP page, contains code for validating the logged-in state based on session and cookie data. For applications that require user authentication, it appears at the beginning of the application page.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:148]: Find a way to put emphasis on file names throughout your article. And again, do this where applicable.

**Review by jumamiller — CHANGES_REQUESTED** [review]
@Juskam, you've an interesting topic. Here are the few issues you should fix:

 - The intro is too wordy, I've suggested a way to resolve it.
 - Explain all your code snippets
 - Remove all unnecessary spaces- I've pointed this ina few sections, fix where applicable.
 - Add the source code github link.
In case of any inquiry, reach me on slack.

**Comment by Juskam** [comment]
@jumamiller,changes rectified

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 6]
```diff
@@ -0,0 +1,356 @@
+### Introduction
+In the Login Script, the 'Remember Me' functionality allows users who have logged in to keep their status. A user's logged-in status is serialized in PHP `sessions`, `cookies`, or other similar storage when they select the `Remember Me` option.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:2]: The paragraph would leave readers scratching their heads. For example, when you say "....allows users who have...to keep their status". Which status are you talking about here?

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+In this edge, you will learn how to create a persistent PHP web application.
+#### Table of contents
+- [Introduction](#introduction)
+- [Preserve logged in state](#create-cookies-to-preserve-the-logged-in-state)
+- [validate remembered login](#validate-remembered-login-with-php-session-and-cookies)
+- [Clear remembered login](#clear-remembered-login-with-session-and-cookies-on-logout)
+- [Controller classes](#auth-and-dbcontroller-classes)
+- [Conclusion](#conclusion)
+- [check the code snippets at github](https://github.com/Juskam/secure-remember-me-for-login-using-php-session-and-cookies)
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:11]: Capitalization mistake!

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+Make sure to look at the source code on GitHub so you know what files to create.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:12]: Do not force readers to visit your Github source code . Cover everything on this tutorial. Github is just a reference for interested readers.

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+### Create cookies to preserve the logged-in state
+To obtain the `username` and `password`, I've created a login form. This form has a 'Remember Me' checkbox for the user's convenience, which allows the user to keep his logged-in state. When a user provides their login information, PHP receives it and compares it to the database of members.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:14]: .....towards the last, "PHP receives it and compares it to database of members....." This sentence could be written in some other better ways. Remember we submit form to server which then processes that data???. database has records? why use `members`?

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+On successful login, if the user selected ‘Remember Me’ then the logged-in status is stored in PHP `session` and `cookies`.
+Because storing the plain password in the cookie is a security flaw, the authentication keys are produced using random integers. These keys are hashed and saved in a database with a one-month expiration time. When the timer runs out, the expiration flag is set to `0` and the keys are turned off.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:16]: Is the expiry restricted to one month?

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+        <div>
+            <input type="submit" name="login" value="Login"
+                class="form-submit-button"></span>
+        </div>
+    </div>
+</form>
+```
+
+### Validate remembered login with PHP session and cookies
+`authCookieSessionValidate.php`, a PHP page, contains code for validating the logged-in state based on session and cookie data. For applications that require user authentication, it appears at the beginning of the application page.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:106]: Rewrite this paragraph. Do not begin with the filename.

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+`$loggedIn` is set to true if it's present in the session or cookie array. It is determined by this boolean value whether or not the user can proceed or be redirected back to the login page.
+First, the PHP session is used to check the remembered login. If it returns false, the function will look in the cookies for the authentication keys. If the keys aren't empty, they'll be hashed and checked against the database.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:108]: .....`checked against the database`. What's a database?

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+        $sql = $this->conn->prepare($query);
+        $this->bindQueryParams($sql, $param_type, $param_value_array);
+        $sql->execute();
+    }
+}
+?>
+```
+### Conclusion
+If you check the "Remember Me" box, your browser will save a cookie so that if you exit the site window without signing out, you will be immediately logged back in the next time you visit. Otherwise, this will not work unless you set your browser to remember cookies.
+If you select the "Remember Me" option, your browser will save a cookie so that if you close the site window without signing out, you will be automatically logged in the next time you visit. Otherwise, unless you set your browser to remember cookies, this will not work.
```
Comment (jumamiller) [inline content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:351]: Conclusion should give us a brief overview of what you covered in the article. So this paragraph should be to moved somewhere else.

**Review by jumamiller — CHANGES_REQUESTED** [review]
@Juskam, I've suggested a few changes. Kindly work on them!

File: content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md (hunk 10 above)
Comment (Juskam) [reply content/articles/php-sessions-and-cookies-to-create-a-secure-remember-me-section-in-a-login-script/Index.md:36]: @jumamiller, from this tutorial I intended to keep users logged in for one month and that explains the restriction to one month

**Comment by hectorkambow** [comment]
@Juskam 
There appear to be many sentences in this article that are taken verbatim from third-party sources. We can't publish any content that is not original or properly cited/referenced. There are many free online tools where you can check this (such as quetext.com, and if you're uncertain about the definitions around plagiarism, this is a [good resource to reference](https://www.scribbr.com/category/plagiarism/.

[PHP Sessions and Cookies to Create a Secure 'Remember Me' Section In a Login Script #3393.pdf](https://github.com/section-engineering-education/engineering-education/files/7027838/PHP.Sessions.and.Cookies.to.Create.a.Secure.Remember.Me.Section.In.a.Login.Script.3393.pdf)
